### PR TITLE
Implement Serialization of SharedArrayBufferObject & $262.agent.* fun…

### PR DESCRIPTION
**Pull request by ksh8281**
…ctions for testing

Signed-off-by: Seonghyun Kim <sh8281.kim@samsung.com>

### CODE REVIEW

File: src/runtime/ArrayBufferObject.cpp [hunk 1]
```diff
@@ -105,7 +105,7 @@ void ArrayBufferObject::detachArrayBuffer()
 
     if (m_backingStore && !m_mayPointsSharedBackingStore) {
         // if backingstore is definitely not shared, we deallocate the backingstore immediately.
-        m_backingStore.value()->deallocate();
+        delete m_backingStore.value();
```
Comment (clover2123) [inline src/runtime/ArrayBufferObject.cpp:108]: Is it safe to delete `BackingStore` directly without `GCFree`?
`BackingStore` is defined as GC-object, so I'm wondering that delete operation is safe for this.

File: src/runtime/SharedArrayBufferObject.cpp [hunk 2]
```diff
@@ -29,6 +29,21 @@
...
+    SharedArrayBufferObjectBackingStoreData* internalData = (SharedArrayBufferObjectBackingStoreData*)deleterData;
+    internalData->deref();
+}
+
+void SharedArrayBufferObjectBackingStoreData::deref()
+{
+    m_refCount--;
+    if (m_refCount == 0) {
+        Global::platform()->onFreeArrayBufferObjectDataBuffer(m_data, m_byteLength);
+        delete this;
```
Comment (clover2123) [inline src/runtime/SharedArrayBufferObject.cpp:43]: `m_refCount` is guaranteed for its atomic operation.
IMHO, but it seems possible that this delete operation could be run simultaneously when several threads go into if statement at the same time (I'm not sure its memory order)
What do you think about it?

File: src/runtime/serialization/Serializer.cpp [hunk 3]
```diff
@@ -65,7 +73,7 @@ bool Serializer::serializeInto(const Value& value, std::ostringstream& output)
     return false;
 }
 
-std::unique_ptr<SerializedValue> Serializer::deserializeFrom(std::istringstream& input)
+std::unique_ptr<SerializedValue> Serializer::deserializeFrom(Context* context, std::istringstream& input)
```
Comment (clover2123) [inline src/runtime/serialization/Serializer.cpp:76]: Parameter `Context* context` seems nowhere used now.

File: src/runtime/ArrayBufferObject.cpp (hunk 1 above)
Comment (ksh8281) [reply src/runtime/ArrayBufferObject.cpp:108]: It is safe.
class gc overrides "operator delete"
```
void gc::operator delete(void* obj) { GC_FREE(obj); }
```
the operator function guarantees memory safety

File: src/runtime/SharedArrayBufferObject.cpp (hunk 2 above)
Comment (ksh8281) [reply src/runtime/SharedArrayBufferObject.cpp:43]: Good point!
I didn't imagine that.
I am going to fix problem right now

File: src/runtime/serialization/Serializer.cpp (hunk 3 above)
Comment (ksh8281) [reply src/runtime/serialization/Serializer.cpp:76]: You are right!
Good point

File: src/shell/Shell.cpp [hunk 4]
```diff
@@ -634,6 +513,390 @@ static bool evalScript(ContextRef* context, StringRef* source, StringRef* srcNam
     return result;
 }
 
+struct WorkerThreadData {
+    std::thread thread;
```
Comment (clover2123) [inline src/shell/Shell.cpp:517]: This `thread` member seems needless too.
each thread is allocated in `workerThreads` vector in the below
```c++
std::vector<std::pair<std::thread, WorkerThreadData>> workerThreads;
```

Comment (ksh8281) [reply src/shell/Shell.cpp]: I removed it!

**Comment by bbrto21** [comment]
This PR break a build with  -DESCARGOT_OUTPUT=shell (not shell_test)

**Comment by ksh8281** [comment]
@bbrto21 Thanks for review. I fixed it

File: src/runtime/SharedArrayBufferObject.cpp (hunk 2 above)
Comment (clover2123) [reply src/runtime/SharedArrayBufferObject.cpp:43]: Looks great 🚀

**Review by clover2123 — APPROVED** [review]
LGTM

**Review by bbrto21 — APPROVED** [review]

**Pull request merged**